src/LinuxTracing/ThreadStateManagerTest.cpp

```diff
@@ -445,6 +471,51 @@ TEST(ThreadStateManager, NoStateChangeWithSwitchIn) {
   EXPECT_EQ(slice->wakeup_reason(), orbit_grpc_protos::ThreadStateSlice::kNotApplicable);
   EXPECT_EQ(slice->wakeup_pid(), kWakeupPidTidWhenWakeupReasonNotApplicable);
   EXPECT_EQ(slice->wakeup_tid(), kWakeupPidTidWhenWakeupReasonNotApplicable);
+  EXPECT_EQ(slice->callstack_status(), orbit_grpc_protos::ThreadStateSlice::kNoCallstack);
+}
+
+TEST(ThreadStateManager, SwitchOutAndWakeupWaitForCallstacks) {
+  constexpr pid_t kTid = 42;
+  constexpr pid_t kWasUnblockedByTid = 420;
+  constexpr pid_t kWasUnblockedByPid = 4200;
+
+  ThreadStateManager manager;
+  std::optional<ThreadStateSlice> slice;
+
+  manager.OnInitialState(100, kTid, ThreadStateSlice::kRunning);
+
+  slice = manager.OnSchedSwitchOut(200, kTid, ThreadStateSlice::kInterruptibleSleep,
+                                   /*wait_for_callstack*/ true);
+  ASSERT_TRUE(slice.has_value());
+  EXPECT_EQ(slice->tid(), kTid);
+  EXPECT_EQ(slice->thread_state(), ThreadStateSlice::kRunning);
+  EXPECT_EQ(slice->duration_ns(), 100);
+  EXPECT_EQ(slice->end_timestamp_ns(), 200);
+  EXPECT_EQ(slice->wakeup_reason(), orbit_grpc_protos::ThreadStateSlice::kNotApplicable);
+  EXPECT_EQ(slice->wakeup_pid(), kWakeupPidTidWhenWakeupReasonNotApplicable);
+  EXPECT_EQ(slice->wakeup_tid(), kWakeupPidTidWhenWakeupReasonNotApplicable);
+  EXPECT_EQ(slice->callstack_status(), orbit_grpc_protos::ThreadStateSlice::kNoCallstack);
+
+  slice = manager.OnSchedWakeup(300, kTid, kWasUnblockedByTid, kWasUnblockedByPid,
+                                /*wait_for_callstack*/ true);
+  EXPECT_EQ(slice->tid(), kTid);
+  EXPECT_EQ(slice->thread_state(), ThreadStateSlice::kInterruptibleSleep);
+  EXPECT_EQ(slice->duration_ns(), 100);
+  EXPECT_EQ(slice->end_timestamp_ns(), 300);
+  EXPECT_EQ(slice->wakeup_reason(), orbit_grpc_protos::ThreadStateSlice::kNotApplicable);
+  EXPECT_EQ(slice->wakeup_pid(), kWakeupPidTidWhenWakeupReasonNotApplicable);
+  EXPECT_EQ(slice->wakeup_tid(), kWakeupPidTidWhenWakeupReasonNotApplicable);
+  EXPECT_EQ(slice->callstack_status(), orbit_grpc_protos::ThreadStateSlice::kWaitingForCallstack);
+
+  slice = manager.OnSchedSwitchIn(400, kTid);
+  EXPECT_EQ(slice->tid(), kTid);
+  EXPECT_EQ(slice->thread_state(), ThreadStateSlice::kRunnable);
+  EXPECT_EQ(slice->duration_ns(), 100);
+  EXPECT_EQ(slice->end_timestamp_ns(), 400);
+  EXPECT_EQ(slice->wakeup_reason(), orbit_grpc_protos::ThreadStateSlice::kUnblocked);
+  EXPECT_EQ(slice->wakeup_pid(), kWasUnblockedByPid);
+  EXPECT_EQ(slice->wakeup_tid(), kWasUnblockedByTid);
+  EXPECT_EQ(slice->callstack_status(), orbit_grpc_protos::ThreadStateSlice::kWaitingForCallstack);
 }
```

Comment:
Can you add another `OnSchedSwitchOut`, since the previous `OnSchedSwitchOut` was following `OnInitialState`, which is a bit different than following `OnSchedSwitchIn`?

Reply:
Done.